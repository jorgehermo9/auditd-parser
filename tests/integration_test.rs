use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
