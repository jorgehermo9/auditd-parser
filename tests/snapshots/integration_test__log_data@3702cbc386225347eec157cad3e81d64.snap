---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/auparse/test3.log:1"
expression: result
info: "node=auditdtest.a1959.org type=SYSCALL msg=audit(1451781471.394:194435): arch=c000003e syscall=23 success=yes exit=1 a0=c a1=56420184ade0 a2=564201867510 a3=0 items=0 ppid=1271 pid=1281 auid=1000 uid=1000 gid=1000 euid=1000 suid=1000 fsuid=1000 egid=1000 sgid=1000 fsgid=1000 tty=(none) ses=1 comm=\"sshd\" exe=\"/usr/sbin/sshd\" subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 key=(null)"
---
{
  "record_type": "SYSCALL",
  "timestamp": 1451781471394,
  "id": 194435,
  "node": "auditdtest.a1959.org",
  "fields": {
    "a0": "c",
    "a1": "56420184ade0",
    "a2": "564201867510",
    "a3": "0",
    "arch": "c000003e",
    "auid": 1000,
    "comm": "sshd",
    "egid": 1000,
    "euid": 1000,
    "exe": "/usr/sbin/sshd",
    "exit": 1,
    "fsgid": 1000,
    "fsuid": 1000,
    "gid": 1000,
    "items": "0",
    "key": "(null)",
    "pid": "1281",
    "ppid": "1271",
    "ses": "1",
    "sgid": 1000,
    "subj": {
      "level": {
        "category": "c0.c1023",
        "sensitivity": "s0-s0"
      },
      "module": "SELinux",
      "role": "unconfined_r",
      "type": "unconfined_t",
      "user": "unconfined_u"
    },
    "success": true,
    "suid": 1000,
    "syscall": "23",
    "tty": "(none)",
    "uid": 1000
  }
}
